data/posts/2017-06-23-expanding-on-phalcon-3-2-0-new-features.md

```diff
@@ -0,0 +1,328 @@
+  use Phalcon\Db\Adapter\Pdo\Factory;
+
+  $di = new Di();
+  $config = new Ini('config.ini');
+  
+  $di->set('config', $config);
+  
+  $di->set('db', function() {
+      return Factory::load($this->config->database);
+  });
+  ```
+  
+  This properly create your database instance and you can change database adapter to other in ini file anytime without changing your code.
+  
+- ##### Added ability to sanitize URL to `Phalcon\Filter`
+  There were added new sanitize filter - `url` which allows you to clear your urls. For example:
+  ```php
+  <?php
+
+  use Phalcon\Filter;
+
+  $filter = new Filter();
+  $wrongUrl = 'http://juhara��.co�m';
+  $correctUrl = $filter->sanitize($wrongUrl, 'url');
+  echo $correctUrl; // displays 'http://juhara.com'
+  ```
+
+- ##### Added `Phalcon\Mvc\Model::hasUpdated` and `Phalcon\Mvc\Model:getUpdatedFields`, way to check if fields were updated after create/save/update, Added option to disable snapshot update on create/save using `Phalcon\Mvc\Model::setup(['updateSnapshotOnSave' => false])` or `phalcon.orm.update_snapshot_on_save = 0` in `php.ini`
+  In Phalcon 3.1.0 we changed behavior of snapshots to be updated on model creation/update. This could potentially cause a problems to your application if you were executing `Model::getChangedFields` in `afterUpdate`, `afterSave` or `afterCreate`. This change was done to fix other things(for example dynamic update). Right now you have two options - change your methods from `hasChanged`, `getChangedFields`, `getSnapshot` to `hasUpdated`, `getUpdatedFiels`, `getOldSnapshot` or add `phalcon.orm.update_snapshot_on_save = 0` to your `php.ini` to disable snapshot updating on save.
+  
+  ```php
+  <?php
+
+  use Phalcon\Mvc\Model;
+
+  class User extends Model
+  {
+      public function initialize()
+      {
+          $this->keepSnapshots(true);
+      }
+  }
+  
+  $user = new User();
+  $user->name = "Test User";
+  $user->create();
+  var_dump($user->getChangedFields());
+  $user->login = "testuser";
+  var_dump($user->getChangedFields());
+  $user->update();
+  var_dump($user->getChangedFields());
+  ```
+  On Phalcon 3.0.0 output was:
+  ```
+  array(1) {
+    [0]=>
+    string(4) "name"
+  }
+  array(2) {
+    [0]=>
+    string(4) "name"
+    [1]=>
+    string(5) "login"
+  }
+  array(2) {
+    [0]=>
+    string(4) "name"
+    [1]=>
+    string(5) "login"
+  }
+  ```
+  
+  On Phalcon 3.1.0 and later it is:
+  ```
+  array(0) {
+  }
+  array(1) {
+    [0]=>
+    string(5) "login"
+  }
+  array(0) {
+  }
+  ```
+  
+  `Model::getUpdatedFields` will properly return updated fields or as mentioned above you can go back to previous behavior by changing ini value.
+  
+- ##### Added support for having option in `Phalcon\Paginator\Adapter\QueryBuilder` [#12111](https://github.com/phalcon/cphalcon/issues/12111)
+  From now you can use `Phalcon\Mvc\Model\Query\Builder::having` and `Phalcon\Paginator\Adapter\QueryBuilder`. Let's assume you have such a table `stock`:
```

Comment:
From now on you can use `Phalcon\Mvc\Model\Query\Builder::having` and `Phalcon\Paginator\Adapter\QueryBuilder`. 

Let's assume you have such a table `stock`: